### PR TITLE
Win screen

### LINKED ISSUE #54: Win screen
show opponent remaining ships

### PR DESCRIPTION
**Pull request by maucon**
see enemy board when game is over
resolves #54

### CODE REVIEW

File: src/main/kotlin/de/twintorx/battleship/game/Ship.kt [hunk 1]
```diff
@@ -21,81 +23,15 @@ enum class Ship(
...
-//                "                   _|_|",
-//                "   __         _____|   |____________",
-//                " >|   \\_____/                       )",
-//                "  |__ ----_________________________/"
-//        )),
-//        DESTROYER(mutableListOf(
-//                "          ___",
-//                "  __--A__/__\\_A--_",
-//                "  \\_______________|"))
-
```
Comment (TheSwerik) [inline src/main/kotlin/de/twintorx/battleship/game/Ship.kt:91]: wtf

**Review by TheSwerik — APPROVED** [review]
This code is [shit](https://youtu.be/dQw4w9WgXcQ)

**Review by DennisGeb — CHANGES_REQUESTED** [review]
- rename params
- move all Writer interaction from Player to Console for clear structure

**Review by DennisGeb — APPROVED** [review]

**Review by DennisGeb — CHANGES_REQUESTED** [review]
see first review

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import de.twintorx.battleship.game.Ship
+import de.twintorx.battleship.game.board.GameBoard
+import de.twintorx.battleship.game.board.TrackBoard
+import de.twintorx.battleship.ui.io.PlayerMessage
+import de.twintorx.battleship.ui.io.Writer
+
+object Console {
+
+    fun printBoards(inGame: Boolean, gameBoard: GameBoard, trackBoard: TrackBoard, rehp: Int, rohp: Int, res: Int, ros: Int) {
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Console.kt:11]: `inGame` -> `isInGame`

wtf is `rehp`, `rohp`, `res`, `ros`

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        printAvailableShips()
+        Writer.println("\n${" " * 4}${PlayerMessage.GAME_BOARD}${" " * (trackBoard.size * 3)}${" " * 7}${PlayerMessage.TRACK_BOARD}")
+        printBoardWithStats(gameBoard, trackBoard, inGame, rehp, rohp, res, ros)
+    }
+
+    fun printWinScreen(won: Boolean, gameBoard: GameBoard, enemyBoard: GameBoard, rehp: Int, rohp: Int, res: Int, ros: Int) {
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Console.kt:17]:
```suggestion
    fun printWinScreen(hasWon: Boolean, gameBoard: GameBoard, enemyBoard: GameBoard, rehp: Int, rohp: Int, res: Int, ros: Int) {
```

again wtf is `rehp`, `rohp`, `res`, `ros`

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 4]
```diff
@@ -62,7 +62,8 @@ class Player {
         Writer.clearConsole()
         val ships = Ship.getStandardShipSet()
         Writer.print("${PlayerMessage.PLACE_SHIPS}\n")
-        printBoards(false)
+        Console.printBoards(false, gameBoard, trackBoard,
+                remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
 
         while (ships.isNotEmpty()) {
             Writer.print("\n${PlayerMessage.CHOOSE_SHIP}\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:69]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 5]
```diff
@@ -79,7 +80,8 @@ class Player {
                 placeShip(this[0]).run {
                     Writer.clearConsole()
                     Writer.print("${PlayerMessage.PLACE_SHIPS}\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:82]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 6]
```diff
@@ -91,7 +93,8 @@ class Player {
         Writer.print("\n${PlayerMessage.WAITING_FOR_PLACEMENT}\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:93]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 7]
```diff
@@ -91,7 +93,8 @@ class Player {
...
         if (client.sendReadyGetTurn().also {
                     Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:95]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 8]
```diff
@@ -132,28 +135,34 @@ class Player {
             Move.HIT -> {
                 Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:136]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 9]
```diff
@@ -132,28 +135,34 @@ class Player {
...
                 remainingEnemyHitPoints--
-                trackBoard.mark(point.x, point.y, Mark.HIT_SHIP).run { printBoards() }
+                trackBoard.mark(point.x, point.y, Mark.HIT_SHIP)
+                Console.printBoards(true, gameBoard, trackBoard,
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
                 Writer.print("\n${PlayerMessage.HIT_SHIP} $shotPosition\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:141]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 10]
```diff
@@ -132,28 +135,34 @@ class Player {
...
                 shoot()
             }
             Move.SUNK -> {
                 Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:145]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 11]
```diff
@@ -132,28 +135,34 @@ class Player {
...
                 remainingEnemyHitPoints--
                 remainingEnemyShips--
-                trackBoard.mark(point.x, point.y, Mark.HIT_SHIP).run { printBoards() }
+                trackBoard.mark(point.x, point.y, Mark.HIT_SHIP)
+                Console.printBoards(true, gameBoard, trackBoard,
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
                 Writer.print("\n${PlayerMessage.SUNK_SHIP} $shotPosition\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:151]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 12]
```diff
@@ -132,28 +135,34 @@ class Player {
...
                 remainingEnemyHitPoints = 0
                 remainingEnemyShips = 0
-                trackBoard.mark(point.x, point.y, Mark.HIT_SHIP).run { printBoards() }
-                Writer.print("\n${PlayerMessage.WIN}\n")
+                Console.printWinScreen(true, gameBoard, client.sendBoardGetBoard(gameBoard),
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
+                client.disconnect()
             }
             Move.NO_HIT -> {
                 Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:162]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 13]
```diff
@@ -132,28 +135,34 @@ class Player {
...
-                trackBoard.mark(point.x, point.y, Mark.HIT_NOTHING).run { printBoards() }
+                trackBoard.mark(point.x, point.y, Mark.HIT_NOTHING)
+                Console.printBoards(true, gameBoard, trackBoard,
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
                 Writer.print("\n${PlayerMessage.HIT_NOTHING} $shotPosition\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:166]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 14]
```diff
@@ -173,35 +182,33 @@ class Player {
         when (move) {
             Move.HIT -> {
                 remainingOwnHitPoints--
-                Writer.clearConsole().also {
-                    printBoards()
-                    Writer.print("\n${PlayerMessage.OPPONENT_HIT} $shotPosition\n")
-                }
+                Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:185]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 15]
```diff
@@ -173,35 +182,33 @@ class Player {
...
                 waitForTurn()
             }
             Move.SUNK -> {
                 remainingOwnHitPoints--
                 remainingOwnShips--
-                Writer.clearConsole().also {
-                    printBoards()
-                    Writer.print("\n${PlayerMessage.OPPONENT_SUNK} $shotPosition\n")
-                }
+                Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:194]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 16]
```diff
@@ -173,35 +182,33 @@ class Player {
...
+                Console.printBoards(true, gameBoard, trackBoard,
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
+                Writer.print("\n${PlayerMessage.OPPONENT_SUNK} $shotPosition\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:197]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 17]
```diff
@@ -173,35 +182,33 @@ class Player {
...
+                Console.printWinScreen(false, gameBoard, client.sendBoardGetBoard(gameBoard),
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
                 client.disconnect()
             }
             Move.NO_HIT -> {
-                Writer.clearConsole().also {
-                    printBoards()
-                    Writer.print("\n${PlayerMessage.OPPONENT_MISSED} $shotPosition\n")
-                }
+                Writer.clearConsole()
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:208]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 18]
```diff
@@ -173,35 +182,33 @@ class Player {
...
+                Console.printBoards(true, gameBoard, trackBoard,
+                        remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
+                Writer.print("\n${PlayerMessage.OPPONENT_MISSED} $shotPosition\n")
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:211]: > move all direct prints to console for clear structure
> -> maybe also move clearConsole stuff from connect method and others to console

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Console.kt:11]:
```suggestion
    fun printBoards(isInGame: Boolean, gameBoard: GameBoard, trackBoard: TrackBoard, remainingEnemyHP: Int, remainingOHP: Int, remainingEnemyShips: Int, remainingOShips: Int) {
```

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Console.kt:17]:
```suggestion
    fun printWinScreen(hasWon: Boolean, gameBoard: GameBoard, enemyBoard: GameBoard, remainingEnemyHP: Int, remainingOwnHP: Int, remainingEnemyShips: Int, remainingOwnShips: Int) {
```

or

```suggestion
    fun printWinScreen(hasWon: Boolean, gameBoard: GameBoard, enemyBoard: GameBoard, remainingEnemyHP: Int, remainingPlayerHP: Int, remainingEnemyShips: Int, remainingPlayerShips: Int) {
```

File: src/main/kotlin/de/twintorx/battleship/ui/Player.kt [hunk 19]
```diff
@@ -216,38 +223,6 @@ class Player {
...
-                    else -> println("${value.first}      ${value.second}")
-                }
-            }
-        }
-    }
-
-    private fun defaultValues(inGame: Boolean, playerMessage: PlayerMessage, color: Color = Color.WHITE, counter: String = ""): String =
-            if (inGame) "$playerMessage ${color.paint(counter)}" else ""
-
     private fun input(msg: String, validationMethod: (String) -> (Boolean) = { true }): String {
```
[the remark this answers is not in the document]
Comment (TheSwerik) [reply src/main/kotlin/de/twintorx/battleship/ui/Player.kt:226]: +1

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+                7 -> Writer.println("${value.first}      ${value.second}     $remainingEnemyShips")
+                13 -> Writer.println("${value.first}      ${value.second}     ${defaultValues(inGame, PlayerMessage.YOU)}")
+                15 -> Writer.println("${value.first}      ${value.second}     $remainingOwnHP")
+                17 -> Writer.println("${value.first}      ${value.second}     $remainingOwnShips")
+                else -> Writer.println("${value.first}      ${value.second}")
+            }
+        }
+    }
+
+
```
Comment (maucon) [inline src/main/kotlin/de/twintorx/battleship/ui/Console.kt:92]: wat

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+        Writer.print("${PlayerMessage.PLACE_SHIPS}\n")
+        printBoards(clearConsole = false, inGame = false, gameBoard = gameBoard, trackBoard = trackBoard,
+                enemyHP = enemyHP, ownHP = ownHP, remEnemyShips = remEnemyShips, remOwnShips = remOwnShips)
+    }
+
+    fun printWaitingForPlacement() {
+        Writer.print("\n${PlayerMessage.WAITING_FOR_PLACEMENT}\n")
+    }
+
+
```
Comment (maucon) [inline src/main/kotlin/de/twintorx/battleship/ui/Console.kt:31]: wat

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+    fun printChooseShip() {
+        Writer.print("\n${PlayerMessage.CHOOSE_SHIP}\n")
+    }
+
+    fun printPlaceShips(gameBoard: GameBoard, trackBoard: TrackBoard, enemyHP: Int, ownHP: Int, remEnemyShips: Int, remOwnShips: Int) {
+        Writer.clearConsole()
+        Writer.print("${PlayerMessage.PLACE_SHIPS}\n")
+        printBoards(clearConsole = false, inGame = false, gameBoard = gameBoard, trackBoard = trackBoard,
+                enemyHP = enemyHP, ownHP = ownHP, remEnemyShips = remEnemyShips, remOwnShips = remOwnShips)
```
Comment (maucon) [inline src/main/kotlin/de/twintorx/battleship/ui/Console.kt:24]: wat the legit fucc

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    fun printWaitingForPlacement() {
+        Writer.print("\n${PlayerMessage.WAITING_FOR_PLACEMENT}\n")
+    }
+
+
+    // prints boards and message for shoot/receive shot
+    fun printShotUpdate(gameBoard: GameBoard, trackBoard: TrackBoard, enemyHP: Int, ownHP: Int, remEnemyShips: Int, remOwnShips: Int, message: String) {
+        printBoards(true, inGame = true, gameBoard = gameBoard, trackBoard = trackBoard, enemyHP = enemyHP, ownHP = ownHP, remEnemyShips = remEnemyShips, remOwnShips = remOwnShips)
```
Comment (maucon) [inline src/main/kotlin/de/twintorx/battleship/ui/Console.kt:34]: waaaaaaaaaaaaaat

File: src/main/kotlin/de/twintorx/battleship/ui/Console.kt [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+                17 -> Writer.println("${value.first}      ${value.second}     $remainingOwnShips")
+                else -> Writer.println("${value.first}      ${value.second}")
+            }
+        }
+    }
+
+
+    fun eraseLastLines(lines: Int) {
+        Writer.eraseLast(lines)
+
```
Comment (maucon) [inline src/main/kotlin/de/twintorx/battleship/ui/Console.kt:95]: watttttt

**Review by DennisGeb — APPROVED** [review]

**Pull request merged**